scylla/src/transport/connection.rs

```diff
@@ -821,33 +829,52 @@ impl Connection {
         }
     }
 
+    fn alloc_stream_id(
+        handler_map: &StdMutex<ResponseHandlerMap>,
+        response_handler: ResponseHandler,
+    ) -> Option<i16> {
+        // We are guaranteed here that handler_map will not be locked
+        // by anybody else, so we can do try_lock().unwrap()
+        let mut lock = handler_map.try_lock().unwrap();
+        if let Some(stream_id) = lock.allocate(response_handler) {
+            Some(stream_id)
+        } else {
+            // TODO: Handle this error better, for now we drop this
+            // request and return an error to the receiver
+            error!("Could not allocate stream id");
+            None
+        }
+    }
+
     async fn writer(
         mut write_half: (impl AsyncWrite + Unpin),
         handler_map: &StdMutex<ResponseHandlerMap>,
         mut task_receiver: mpsc::Receiver<Task>,
     ) -> Result<(), QueryError> {
+        const OUTPUT_BUFFER_MAX: usize = 8192;
         // When the Connection object is dropped, the sender half
         // of the channel will be dropped, this task will return an error
         // and the whole worker will be stopped
-        while let Some(task) = task_receiver.recv().await {
-            let stream_id = {
-                // We are guaranteed here that handler_map will not be locked
-                // by anybody else, so we can do try_lock().unwrap()
-                let mut lock = handler_map.try_lock().unwrap();
-
-                if let Some(stream_id) = lock.allocate(task.response_handler) {
-                    stream_id
-                } else {
-                    // TODO: Handle this error better, for now we drop this
-                    // request and return an error to the receiver
-                    error!("Could not allocate stream id");
-                    continue;
+        while let Some(mut task) = task_receiver.recv().await {
+            let mut num_requests = 0;
+            let mut total_sent = 0;
+            while let Some(stream_id) = Self::alloc_stream_id(handler_map, task.response_handler) {
+                let mut req = task.serialized_request;
+                req.set_stream(stream_id);
+                let req_data: &[u8] = req.get_data();
+                total_sent += req_data.len();
+                num_requests += 1;
+                write_half.write_all(req_data).await?;
+                if total_sent >= OUTPUT_BUFFER_MAX {
+                    break;
                 }
```

Comment:
I think this `total_sent >= OUTPUT_BUFFER_MAX` check can be removed as it might make things less efficient. If there is not enough space to place the current request into the buffer (which will happen when `total_sent > OUTPUT_BUFFER_MAX`), then the `write_all` highlighted here will pass buffered data to the underlying writer - which translates to a syscall - and then it will place the current frame into the buffer. Because you `break` in such case, you will soon call `flush` which causes another syscall for the single frame. I think it's better not to break because there may be more requests on the `task_receiver` channel and we will benefit from sending them in a single syscall with the current frame.